File: python/src/WaveletTree.py
import StringIO

from WaveletNode import WaveletNode

"""
    WaveletTree class represents one Wavelet tree data structure
"""
class WaveletTree:
    def __init__(self):
        self.root_node = None

    """ Constracting WaveletTree based on 'data' and 'alphabet' """
    def build(self, alphabet, data):
        self.root_node = WaveletNode(alphabet)
        self.__build(self.root_node, alphabet, data)

    """ Counts number of times 'character' appears in string up to 'position' """
    def rank(self, position, character):
        self.check_position(position)
        self.check_character(character)

        # TODO it will not count character at given 'position'. Does it need to count it?
        return WaveletTree.__rank(self.root_node, position+1, character)

    """ Returns character at 'index' position in string """
    def access(self, index):
        self.check_position(index)

        return WaveletTree.__access(self.root_node, index)

    """ Returns position in string at which 'character' occurs for the 'nth_occurrence' time """
    def select(self, nth_occurrence, character):
        self.check_character(character)

        # Find node that represents 'character'
        tmp_node = self.root_node
        while len(tmp_node.alphabet) != 1:
            is_left = character in tmp_node.left.alphabet
            tmp_node = tmp_node.left if is_left else tmp_node.right

        return WaveletTree.__select(tmp_node.parent, nth_occurrence, character)

    def log(self):
        WaveletTree.__log(self.root_node)

    def check_position(self, position):
        if position > len(self.root_node.bit_vector):
            raise ValueError('Position is greater than bit_vector')

    def check_character(self, character):
        if character not in self.root_node.alphabet:
            raise ValueError('Character is not in alphabet')

    @staticmethod
    # Internal recursive method that builds tree
    def __build(node, alphabet, data):
        if len(alphabet) in [0, 1]:
            return

        half = len(alphabet) / 2

        # Find left && right alphabet
        left_sub_alphabet = alphabet[:half]
        right_sub_alphabet = alphabet[half:]

        # Build bit vector for current node
        buffer = []
        [buffer.append('0' if c in left_sub_alphabet else '1') for c in data]
        node.add(''.join(buffer))

        # Split data for left and right node
        left_sub_data = []
        right_sub_data = []
        for bit in data:
            if bit in left_sub_alphabet:
                left_sub_data.append(bit)
            else:
                right_sub_data.append(bit)

        # Create left && right node and build recursively
        node.left = WaveletNode(left_sub_alphabet, node)
        WaveletTree.__build(node.left, left_sub_alphabet, left_sub_data)
        node.right = WaveletNode(right_sub_alphabet, node)
        WaveletTree.__build(node.right, right_sub_alphabet, right_sub_data)

    @staticmethod
    # Internal recursive method that executes rank query
    def __rank(node, position, character):
        if len(node.alphabet) == 1:
            return position

        is_left = character in node.left.alphabet

        child = node.left if is_left else node.right
        bit_type = '0' if is_left else '1'

        new_position = 0
        for bit in node.bit_vector[:position]:
            if bit == bit_type:
                new_position += 1

        return WaveletTree.__rank(child, new_position, character)

    @staticmethod
    # Internal recursive method that executes access query
    def __access(node, index):
        if len(node.alphabet) == 1:
            return node.alphabet[0]

        is_left = node.bit_vector[index] == '0'
        child = node.left if is_left else node.right
        bit_type = '0' if is_left else '1'

        new_index = 0
        for bit in node.bit_vector[:index]:
            if bit == bit_type:
                new_index += 1

        return WaveletTree.__access(child, new_index)


    @staticmethod
    # Internal recursive method that executes select query
    def __select(node, nth_occurence, character):
        is_left = character in node.left.alphabet
        bit_type = '0' if is_left else '1'

        new_index = node.nth_occurence(nth_occurence, bit_type)

        return new_index if not node.parent else WaveletTree.__select(node.parent, new_index + 1, character)

    @staticmethod
    # Internal recursive helper method for logging node and all its childs
    def __log(node):
        if not node or len(node.bit_vector) == 0:
            return

        print node.bit_vector
        WaveletTree.__log(node.left)
        WaveletTree.__log(node.right)

File: Java/src/WaveletTree.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Set;
import java.util.TreeSet;

/**
 * Wavelet tree implementation.
 * 
 * @author ICapalija
 * 
 */
public class WaveletTree {

	/**
	 * Root node of the tree.
	 */
	private WaveletNode rootNode;
	/**
	 * Sequence alphabet.
	 */
	private byte[] alphabet;

	/**
	 * Wavelet tree constructor.
	 * 
	 * @param sequence
	 *            character sequence.
	 */
	public WaveletTree(byte[] sequence) {
		// convert sequence to list
		List<Byte> seq = new ArrayList<>();
		for (byte l : sequence) {
			seq.add(l);
		}
		
		this.alphabet = getDistinct(seq);
		System.out.println(Arrays.toString(this.alphabet));
		this.rootNode = new WaveletNode(seq.size(), (byte) 0,
				(byte) (this.alphabet.length - 1));

		// build tree
		this.buildTree(this.rootNode, seq, 0, this.alphabet);
	}

	/**
	 * Build tree.
	 * 
	 * @param node
	 *            current node.
	 * @param sequence
	 *            char sequence.
	 * @param currDepth
	 *            tree depth.
	 * @param alphabet
	 *            sequence alphabet.
	 */
	private void buildTree(WaveletNode node, List<Byte> sequence,
			int currDepth, byte[] alphabet) {
		// current node alphabet size
		byte alphSize = (byte) (node.getEndAlphabetIndex()
				- node.getStartAlphabetIndex() + 1);
		// current node alphabet split
		byte alphHalf = (byte) (node.getStartAlphabetIndex() + alphSize / 2);

		List<Byte> leftSeq = new ArrayList<>();
		List<Byte> rightSeq = new ArrayList<>();

		// encode letters into bits
		for (int i = 0; i < sequence.size(); i++) {
			byte letter = sequence.get(i);

			if (letter < alphabet[alphHalf]) {// if the letter is on the left
												// side of the alphabet
				leftSeq.add(letter);
				node.setBit(i, false);
			} else {// if the letter is on the right side of the alphabet
				rightSeq.add(letter);
				node.setBit(i, true);
			}
		}

		// if left child's alphabet is greater than 0
		if ((alphHalf - node.getStartAlphabetIndex() - 1) > 0) {
			node.setLeftChild(new WaveletNode(leftSeq.size(), node
					.getStartAlphabetIndex(), (byte) (alphHalf - 1), node));
			buildTree(node.getLeftChild(), leftSeq, currDepth + 1, alphabet);
		}

		// if right child's alphabet is greater than 0
		if ((node.getEndAlphabetIndex() - alphHalf) > 0) {
			node.setRightChild(new WaveletNode(rightSeq.size(), alphHalf, node
					.getEndAlphabetIndex(), node));
			buildTree(node.getRightChild(), rightSeq, currDepth + 1, alphabet);
		}
	}

	/**
	 * Rank method. 
	 * Rank is number of occurrences of given letter
	 * in prefix S[0..index].
	 * 
	 * @param letter
	 *            letter whose rank is calculated.
	 * @param index
	 *            end index of prefix.
	 * @return letter rank in prefix.
	 */
	public int rank(byte letter, int index) {
		return rank(this.rootNode, index + 1, letter);
	}

	/**
	 * Recursive method that calculates rank. Rank is number of 
	 * occurrences of given letter in prefix S[0..index].
	 * 
	 * @param node
	 *            current node.
	 * @param index
	 *            end index of prefix.
	 * @param letter
	 *            letter whose rank is calculated.
	 * @return letter rank in prefix.
	 */
	private int rank(WaveletNode node, int index, byte letter) {
		byte alphSize = (byte) (node.getEndAlphabetIndex()
				- node.getStartAlphabetIndex() + 1);
		byte alphHalf = (byte) (node.getStartAlphabetIndex() + alphSize / 2);

		if (letter < this.alphabet[alphHalf]) {// if letter is 0 encoded
			// count number of occurrences of 0 in bit array
			int falseCount = node.countOccurrence(false, index);

			if (node.getLeftChild() == null) {// if node is leaf
				return falseCount;
			} else {// if node is not leaf
				return rank(node.getLeftChild(), falseCount, letter);
			}
		} else {// if letter is 1 encoded
			// count number of occurrences of 1 in bit array
			int trueCount = node.countOccurrence(true, index);

			if (node.getRightChild() == null) {// if node is leaf
				return trueCount;
			} else {// if node is not leaf
				return rank(node.getRightChild(), trueCount, letter);
			}
		}
	}

	/**
	 * Select method.
	 * Result of a select method is smallest prefix with 
	 * <code>occurrenceNum + 1</code> occurrences of given letter.
	 * 
	 * @param letter
	 *            letter.
	 * @param occurrenceNum
	 *            number of occurrences in prefix.
	 * @return end index of result prefix.
	 */
	public int select(byte letter, int occurrenceNum) {
		// calculate index of the letter in alphabet, which is starting point of
		// select method
		int letterIndex = Arrays.binarySearch(this.alphabet, letter);
		WaveletNode currNode = this.rootNode;

		// find starting point node
		while (true) {
			byte alphSize = (byte) (currNode.getEndAlphabetIndex()
					- currNode.getStartAlphabetIndex() + 1);
			byte alphHalf = (byte) (currNode.getStartAlphabetIndex() + alphSize / 2);

			if (alphHalf > letterIndex) {
				if (currNode.getLeftChild() == null) break;
				currNode = currNode.getLeftChild();
			} else {
				if (currNode.getRightChild() == null) break;
				currNode = currNode.getRightChild();
			}
		}
		

		// call recursive method
		return select(currNode, letter, occurrenceNum, alphabet);
	}

	/**
	 * Recursive calculation of select method.
	 * 
	 * @param node
	 *            current node.
	 * @param letter
	 *            letter.
	 * @param occurrenceNum
	 *            number of occurrences in prefix.
	 * @param alphabet
	 *            alphabet.
	 * @return end index of result prefix.
	 */
	private int select(WaveletNode node, byte letter, int occurrenceNum,
			byte[] alphabet) {
		byte alphSize = (byte) (node.getEndAlphabetIndex()
				- node.getStartAlphabetIndex() + 1);
		byte alphHalf = (byte) (node.getStartAlphabetIndex() + alphSize / 2);

		// is letter 0 or 1 encoded
		boolean bitType = letter >= alphabet[alphHalf];

		// calculate length of prefix of an array which contains
		// <code>occurranceNum</code> bits of type <code>bitType</code>
		int index = node.calcOccurrenceArrayLength(bitType, occurrenceNum + 1);

		if (node.getParent() == null) {// if root return result
			return index;
		} else {// else ask parent
			return select(node.getParent(), letter, index, alphabet);
		}
	}

	/**
	 * Access method.
	 * Get character on given index in sequence.
	 * 
	 * @param index
	 *            index of character.
	 * @return character on given index.
	 */
	public byte access(int index) {
		return access(this.rootNode, index, this.alphabet);
	}

	/**
	 * Get character on given index in sequence.
	 * 
	 * @param node
	 *            current node.
	 * @param index
	 *            index of character.
	 * @param alphabet
	 *            alphabet.
	 * @return character on given index.
	 */
	public byte access(WaveletNode node, int index, byte[] alphabet) {

		if (!node.getBit(index)) {// if letter is 0 encoded

			if (node.getLeftChild() == null) {// if node is leaf
				return alphabet[node.getStartAlphabetIndex()];
			} else {// if node is not leaf
				// count number of occurrences of 0 in bit array
				int falseCount = node.countOccurrence(false, index);
				return access(node.getLeftChild(), falseCount, alphabet);
			}

		} else {// if letter is 1 encoded

			if (node.getRightChild() == null) {// if node is leaf
				return alphabet[node.getEndAlphabetIndex()];
			} else {// if node is not leaf
				// count number of occurrences of 1 in bit array
				int trueCount = node.countOccurrence(true, index);
				return access(node.getRightChild(), trueCount, alphabet);

			}

		}
	}

	/**
	 * Convert tree to string.
	 * 
	 * @return string representation of tree.
	 */
	public String toString() {
		Queue<WaveletNode> queue = new LinkedList<>();
		WaveletNode temp;
		StringBuilder builder = new StringBuilder();

		queue.add(this.rootNode);
		int counter = 1;
		while (!queue.isEmpty()) {
			counter++;

			// remove the node from the queue
			temp = queue.poll();

			// process current node
			builder.append(temp.toString());
			builder.append(' ');
			double pom = Math.log(counter) / Math.log(2);
			if (pom == (int) pom) {
				builder.append('\n');
			}

			// process the left child first if not null
			if (temp.getLeftChild() != null) {
				queue.add(temp.getLeftChild());
			}
			// process the right child after the left if not null
			if (temp.getRightChild() != null) {
				queue.add(temp.getRightChild());
			}
		}

		return builder.toString();
	}

	/**
	 * Get distinct values from array.
	 * 
	 * @param sequence
	 *            sequence.
	 * @return byte of sorted distinct values.
	 */
	private byte[] getDistinct(List<Byte> sequence) {
		// get distinct values
		Set<Byte> set = new TreeSet<Byte>(sequence);

		// convert to byte array
		byte[] distinct = new byte[set.size()];
		int index = 0;
		for (byte letter : set) {
			distinct[index] = letter;
			index++;
		}

		return distinct;
	}

	/**
	 * Calculate memory usage.
	 * 
	 * @return memory usage.
	 */
	public int getMemoryUsage() {
		return 4 + this.alphabet.length + this.rootNode.getMemoryUsage();
	}

}
